packages/py-moose-lib/moose_lib/streaming_runner.py

```diff
@@ -0,0 +1,331 @@
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+)
+logger = logging.getLogger("moose-streaming-runner")
+
+class StreamingRunner:
+    """
+    Main class for running streaming transformations in Moose Python projects.
+    
+    This handles the lifecycle of consuming messages, applying transformations,
+    and producing output messages.
+    """
+    
+    def __init__(self, 
+                 source_topic: str,
+                 target_topic: Optional[str],
+                 broker: str,
+                 transform_function: Callable,
+                 max_subscriber_count: int = 1,
+                 sasl_username: Optional[str] = None,
+                 sasl_password: Optional[str] = None,
+                 sasl_mechanism: Optional[str] = None,
+                 security_protocol: Optional[str] = None):
+        """
+        Initialize the streaming runner.
+        
+        Args:
+            source_topic: Name of the source Kafka topic
+            target_topic: Name of the target Kafka topic (optional for multi-destination)
+            broker: Kafka broker address
+            transform_function: Function to transform the data
+            max_subscriber_count: Maximum number of subscribers
+            sasl_username: SASL username for authentication
+            sasl_password: SASL password for authentication
+            sasl_mechanism: SASL mechanism for authentication
+            security_protocol: Security protocol
+        """
+        if not KAFKA_AVAILABLE:
+            raise ImportError(
+                "kafka-python-ng is required for streaming functions. "
+                "Install it with: pip install kafka-python-ng"
+            )
+            
+        self.source_topic = source_topic
+        self.target_topic = target_topic
+        self.broker = broker
+        self.transform_function = transform_function
+        self.max_subscriber_count = max_subscriber_count
+        
+        # Authentication config
+        self.sasl_config = None
+        if sasl_username and sasl_password and sasl_mechanism:
+            self.sasl_config = {
+                'sasl_plain_username': sasl_username,
+                'sasl_plain_password': sasl_password,
+                'sasl_mechanism': sasl_mechanism,
+                'security_protocol': security_protocol or 'SASL_PLAINTEXT'
+            }
+            
+        # Metrics
+        self.metrics = {
+            'count_in': 0,
+            'count_out': 0,
+            'bytes_processed': 0
+        }
+        
+        # Thread management
+        self._running = False
+        self._threads = []
+        
+    def _create_consumer(self) -> KafkaConsumer:
+        """Create and configure a Kafka consumer."""
+        config = {
+            'bootstrap_servers': self.broker,
+            'group_id': f'moose-streaming-{self.source_topic}-{self.target_topic or "multi"}',
```

Comment:
iirc, there's an assumption for redpanda acls, it expects a certain format for the group id so if we change this, infra likely has to change as well.